### PR TITLE
fix nesting volumes even more deeply

### PR DESCRIPTION
**Pull request by vito**
the previous fix actually only made it work slightly more deeply; this
should fix it all the way. the problem was that the grandparent was
found using its logical parent, rather than the 'actual' parent, which
must be found by recursively walking through the hierarchy all the way
to the bottom.

* test up to depth 10
* assert that file updates and deletions are preserved at each depth,
  just to be sure

Signed-off-by: Alex Suraci <suraci.alex@gmail.com>

### CODE REVIEW

**Comment by muntac** [comment]
@vito the recursive logic seems sensible.

I know the tests are passing on the CI but wanted to run the tests myself to get a better understanding of what's going on. I'm running the test on the `concourse/baggageclaim-ci` image and using the `unit-linux` script similar to CI.

I run into 2 issues:
1. On both `master` and this branch the tests fails in `overlay_linux_test.go` on this [line](https://github.com/concourse/baggageclaim/blob/362da05c12f106a0eef44f1137f00c5a3227ff31/volume/driver/overlay_linux_test.go#L58)
```
childVolInit, err := rootVolLive.NewSubvolume("child-vol")
Expect(err).ToNot(HaveOccurred())
```

Error:
```
• Failure [0.008 seconds]
Overlay Driver [It] supports nesting >2 levels deep
/baggageclaim/volume/driver/overlay_linux_test.go:37

  Unexpected error:
      <syscall.Errno>: 0x16
      invalid argument
  occurred

  /baggageclaim/volume/driver/overlay_linux_test.go:59
```

2. To run the `unit-lixux` script properly I have to run `mkdir /scratch` first as it complains about not having scratch at this [line](https://github.com/concourse/baggageclaim/blob/362da05c12f106a0eef44f1137f00c5a3227ff31/ci/scripts/unit-linux#L46)

Are you able to successfully run the tests locally?

**Comment by vito** [comment]
@muntac Running the tests locally is a bit of a pain because you have to run them as `root`. I was able to run them all locally though and they passed, so not sure what's up. I wouldn't use `ci/scripts/unit-linux` though - it assumes it's running in a Concourse task and does a bunch of loopback device setup for the `btrfs` tests too. Pretty sure I just did `ginkgo -r -p` in a `root` dev setup.

**Review by muntac — APPROVED** [review]

**Comment by vito** [comment]
@muntac Figured out why you were seeing `invalid argument`! This commit explains it: https://github.com/concourse/baggageclaim/pull/43/commits/d6ac0b3c27ed429ef57fbb15c001ef92dca4a10a

tl;dr: it would pass if the image was fetched on another worker and streamed to the worker running the tests. it would fail if the task ran with a copy-on-write of a local cache of the image.
